imageio/plugins/pillow.py

```diff
@@ -4,874 +4,278 @@
-            nchannels = len(mode)
-            # Shape it.
-            p.shape = -1, nchannels
-            if p.shape[1] == 3 or (p.shape[1] == 4 and mode[-1] == "X"):
-                p = np.column_stack((p[:, :3], 255 * np.ones(p.shape[0], p.dtype)))
-            # Swap the axes if the mode is in BGR and not RGB
-            if mode.startswith("BGR"):
-                p = p[:, [2, 1, 0]] if p.shape[1] == 3 else p[:, [2, 1, 0, 3]]
-            # Apply palette
-            frame_paletted = np.array(im, np.uint8)
-            try:
-                frame = p[frame_paletted]
-            except Exception:
-                # Ok, let PIL do it. The introduction of the branch that
-                # tests `im.info['transparency']` should make this happen
-                # much less often, but let's keep it, to be safe.
-                frame = im.convert("RGBA")
-        else:
-            # Let Pillow do it. Unlinke skimage, we always convert
-            # to RGBA; palettes can be RGBA.
-            if True:  # im.format == 'PNG' and 'transparency' in im.info:
-                frame = im.convert("RGBA")
-            else:
-                frame = im.convert("RGB")
-    elif "A" in im.mode:
-        frame = im.convert("RGBA")
-    elif im.mode == "CMYK":
-        frame = im.convert("RGB")
-
-    # Apply a post-convert if necessary
-    if as_gray:
-        frame = frame.convert("F")  # Scipy compat
-    elif not isinstance(frame, np.ndarray) and frame.mode == "1":
-        # Workaround for crash in PIL. When im is 1-bit, the call array(im)
-        # can cause a segfault, or generate garbage. See
-        # https://github.com/scipy/scipy/issues/2138 and
-        # https://github.com/python-pillow/Pillow/issues/350.
-        #
-        # This converts im from a 1-bit image to an 8-bit image.
-        frame = frame.convert("L")
-
-    # Convert to numpy array
-    if im.mode.startswith("I;16"):
-        # e.g. in16 PNG's
-        shape = im.size
-        dtype = ">u2" if im.mode.endswith("B") else "<u2"
-        if "S" in im.mode:
-            dtype = dtype.replace("u", "i")
-        frame = np.frombuffer(frame.tobytes(), dtype).copy()
-        frame.shape = shape[::-1]
-    else:
-        # Use uint16 for PNG's in mode I
-        if im.format == "PNG" and im.mode == "I" and dtype is None:
-            dtype = "uint16"
-        frame = np.array(frame, dtype=dtype)
-
-    return frame
-
-
-def ndarray_to_pil(arr, format_str=None, prefer_uint8=True):
-
-    from PIL import Image
-
-    if arr.ndim == 3:
-        arr = image_as_uint(arr, bitdepth=8)
-        mode = {3: "RGB", 4: "RGBA"}[arr.shape[2]]
-
-    elif format_str in ["png", "PNG"]:
-        mode = "I;16"
-        mode_base = "I"
-
-        if arr.dtype.kind == "f":
-            arr = image_as_uint(arr)
-
-        elif prefer_uint8 and arr.max() < 256 and arr.min() >= 0:
-            arr = arr.astype(np.uint8)
-            mode = mode_base = "L"
-
+            except UnidentifiedImageError:
```

Comment:
I think we can simply use the code from can_read and can_write here. As it is now, every image is opened twice.

Also, let's raise a more specific error when a plugin cannot process a request. `NotImplementedError` seems suited, or perhaps a custom error based on it.

Reply:
I ended up going with an exception instead of can_read/can_write, because you mentioned in a comment that plugins should raise an exception if they can't handle a request ( https://github.com/imageio/imageio/pull/574#issuecomment-810220988 ). It follows the pythonic mantra of "better ask for forgiveness than for permission". Personally, I like the idea from an API design perspective: A plugin just handles the request, and if that fails it screams .... _loudly_.

> As it is now, every image is opened twice.

Maybe I misunderstood something, but in my mind the image file is handled by the stateful request object and will be kept open. We hand the same request to every plugin and, once bytes are read, they will remain in a memory buffer; this should be very fast to read.

> Also, let's raise a more specific error when a plugin cannot process a request. NotImplementedError seems suited, or perhaps a custom error based on it.

I was thinking about this, too, when I was writing that code. 

I first had a `NotImplementedError`, but I think doesn't match what is happening. It's not that the code to read/write/do-stuff isn't implemented - chances are it will _never_ be implemented -, but rather that the request can not be handled (e.g. we are trying to open a `.mp4` using `pillow`). Essentially, the input value passed to the constructor function is not valid, so I ended up choosing `ValueError`. I'm happy to raise a different error though if you think it makes more sense.

I have a personal thing against custom errors. I've never used them in a real project, and when I've encountered them they could have likely been refactored into a standard error + a good error string/message. We can introduce them, but I don't really understand how they add anything over `StandardError("This is what went wrong.")`.

Reply:
Ah, I just realized I may have misunderstood you. Do you mean using legacy_pillow's can_read/can_write to check if pillow is compatible?

Reply:
Sorry for not being more clear. I meant using the *code* in the old canread/canwrite to better deterine when to raise an error here :)

> Maybe I misunderstood something, but in my mind the image file is handled by the stateful request object and will be kept open.

Yes you're right. I was triggered by the use of `get_file`. Ideally the code that determines whether or not a file can be dealt with should only use `getfirstbytes`. Think e.g. of non-seekable files.

In general I am opposed to custom errors too. But in this case we really want to distinguish the plugins I-cannot-handle-this signal from something actually going wrong. E.g. with the `ValueError` now, it's very easy for a bug in the plugin to raise ValueError and the format manager to therefore consider that the plugin cannot deal with the request. And the user will never know.

Reply:
Regarding `get_file`, I can use `can_read`/`can_write` code instead. It feels a bit like we are digging in `pillow`'s internals here, but I will see if I can refactor things.

---

Regarding the error, I'll change it to a custom error and we can see how things play out.

Reply:
@almarklein I introduced `InitializationError` as a custom error to catch instead of `ValueError`.

---

I looked through the code of `_can_read` and `_can_write` and I would like to optimize/change it in a way that it no longer relies on hardcoded formats to choose how to initialize

https://github.com/imageio/imageio/blob/6fbab81e7598af847c840c93c9ebec9a94d1f242/imageio/plugins/pillow.py#L97

Instead, I want it to run `Image.preinit()`, check if the format can be supported, and if not run `Image.init()` and try again. This way, we automatically match pillow's behavior if they change the formats included in `preinit()`. (For example, `.tiff` used to be pre-initialized, but that is no longer the case). This is also what `Image.open` does under the hood.

The second check we do in _can_read/_can_write is to check if our plugin has a matching, loaded plugin inside pillow 

https://github.com/imageio/imageio/blob/6fbab81e7598af847c840c93c9ebec9a94d1f242/imageio/plugins/pillow.py#L106

https://github.com/imageio/imageio/blob/6fbab81e7598af847c840c93c9ebec9a94d1f242/imageio/plugins/pillow.py#L123

In the current implementation, this is the only place where we need knowledge of available pillow plugins; which would be a priori knowledge, since we need to create the correct id string. It will be problematic if a user is using custom pillow plugins, since we wouldn't be able to detect them. A better way would be to iterate over all currently loaded formats and see if they can support our file/uri. This (+ some extra legwork) is again what `Image.open(...)` does under the hood.

Another problem here is that our current _can_read/_can_write do not respect pillow's `formats=...` option. The option essentially allows excluding some pillow plugins, and if a user passes `imopen("some.png", plugin="pillow", formats=["PNG", "JPEG"]` I would like to do the expected thing and honor that if possible. I am happy to update the current implementation, but this is again a dublication of pillow's `Image.open`.

This brings me to the crux of the problem, which is our desire to read `getfirstbytes` once and reuse it for each of pillow's plugins. Internally, `Image.open` does this already; it reads the beginning of the file, stores it, and then passes that to each of it's plugins. The only gain I can see from using `getfirstbytes` instead is that we don't have to read the first 16 bytes if pillow wasn't the first backend to be tested. Somehow I doubt this will be a performance bottleneck.

The other topic is non-seekable files, because pillow expects a `seek`able object. For this, our Request object wraps any non-seekable file into a `Request.SeekableFileObject`, which essentially adds an in-memory cache to make the object seekable and available to other backends later on.

So the bottom line, regarding can_read, is that - from my current perspective - it duplicates most of the current Image.open. That said, it could be that I am missing something here.

Regarding _can_write, I agree that it can be useful to check the supported formats upon initialization and fail if there is no compatible one. I wonder if pillow can already do this for us. I will investigate more.

Reply:
Ok, so perhaps its safer (and easier) to use `Image.open` and if needed rewind the file before passing it to a next plugin. Perhaps we can replace the whole firstbytes thing with a file object with a cache (eventualy).

Reply:
> and if needed rewind the file before passing it to a next plugin

That is a very good point. Would it be a good idea for `imopen` to do a `seek(0)` defensively before passing the request to a new plugin while searching?

> Perhaps we can replace the whole firstbytes thing with a file object with a cache (eventualy).

That thought may have crossed my mind as well a few times while I was going through the code. Since we will do some refactoring on the plugins anyway, we can depreciate `firstbytes` in the process?

Reply:
Sure let's have a try.

About the rewinding, if the user has given a file object, in theory the image may not start at pos 0. So we should `tell()` before doing anything, and rewind to that position.

Reply:
> So we should tell() before doing anything, and rewind to that position.

This smells funny. What if the backend (unaware that it is reading a file object) decides to `seek(0)` in the process of reading header data? It sounds like that would completely break things.

Reply:
True. Backends woul also have to take into account this possibility. Maybe this envisioned use-case is just too rare to think about. Let's just do `seek(0)` for now. Maybe add a note there.

Reply:
I checked the code for saving images in pillow and there it makes sense to keep logic along the lines of `_can_write`, because we would like to fail plugin creation before the user actually calls `plugin.save`. I hope I will get time to do implementations again soon :)

---

On a side note, I noticed that the API design is a little flawed (every design is somehow) in that it will inevitably open the file during plugin construction to identify the plugin that needs to be loaded. This is mildly annoying, because it happens outside the context manager, which could leak resources if the program crashes between instantiation of the plugin and creation of the context.

I think for now this is something we can live with (it was like this before), but in the long run, I don't know what the best way to solve it is. We could return a `LazyPlugin`, where common initialization is done, and then - on the first method call, which should happen inside a context manager - it evolves into a real plugin or raises an exception.

An alternative (may be easier) is to simply closes the file when we are done initializing and re-open it inside the context manager. This is pretty simple, but opens the file twice, which isn't ideal.

In either case, I think this is something for a new PR to fix

Reply:
> in that it will inevitably open the file during plugin construction to identify the plugin that needs to be loaded. 

Can't we simply detect errors during the plugin selection process and close the file if needed?

Reply:
Yep, of course imopen can cleanup after itself if it can't find a plugin.  Its more so that

```python
import imageio.v3 as iio
plugin_instance = iio. imopen("file_name.jpg")
assert False
plugin_instance.read(...)
```

would be valid, but might not properly close the file as the `assert False` happens outside of a context manager. Of course, we can overwrite `__del__` in each plugin to handle destruction. THat could be another solution.

Reply:
@almarklein I was working on the `seek(0)` we discussed and noticed that `Request.py` checks that a file exists by simply checking if the given path string exists which could obviously be a dictionary, symlink, or something else (line 375ff). 

Normally, I would treat this as a bug that should be fixed, but there is a comment `# Reading: check that the file exists (but is allowed a dir)`. Do we depend on Request allowing directories as valid paths to read from? I'm asking, because it has some weird consequences for some of our tests, e.g., on windows it allows imopen(""), but then later fails on `instance.read()` with a permission error (that falls a bit out of the sky).

Reply:
mmm ... there are some plugins which support reading a bunch of files from a directory (e.g. DICOM, could be the only one). Maybe we should consider making this more strict and allowing another solution for the dicom plugin.

Reply:
ok, in this case I will leave the `seek(0)` behavior for another PR. I tried it just now and it looks like I need to dig into FreeImage and DICOM to see why their tests break if I reset the file before handing it over to `LegacyPlugin`. I would want to do this in a new PR that is a little more reviewable.

Reply:
I've added the `_can_write` logic to pillow's constructor. There is one change I need to make: the `elif` should be an `else`. Afterward coverage should go up to 100% for pillow again. Otherwise, the changes of today should conclude this thread as well.

This seems to be the last open point for this PR. If there is anything else from your side @almarklein please feel free to comment, otherwise I'd say we summarize in the main thread and then get this thing merged :)

Reply:
I forgot where we are now on the rewinding/first_bytes/auto-caching stuff, but maybe we should just cover that in subsequent prs :)

Reply:
Anything else you want to do / summarize before I hit that big green button?

Reply:
> I forgot where we are now on the rewinding/first_bytes/auto-caching stuff, but maybe we should just cover that in subsequent prs :)

Yes, that's why I don't want to add more stuff here and start adding smaller chunks in dedicated PRs.

My last stand is that the best approach would be to refactor Request to have sole ownership of the raw file. It reads the file into memory on demand (maybe we can use python's mmap module here) and upon `get_file()` it serves a new read-only file object to the plugin which is destroyed once the plugin is done with it. In there, plugins get to do what they want and we can optimize serving cached content vs content read on-demand based on file size and and access frequency, etc.